labs/LanguageBasics/LanguageBasics.md

```diff
@@ -2,39 +2,281 @@
+base = 10` to see numbers in base 10.
+
+
+Comments
+--------
 
 * `//` to end of line
 * `/*` ... `*/` block comment
 
-## The Types of Functions
 
-* The Cryptol interpreter command `:t` is very useful for helping you understand types. For instance the type of the `abs` function which we will define later is displayed by:
+Identifiers
+-----------
+
+Cryptol idenfiers consist of alphanumeric characters plus `'`
+(apostrophe, but read "prime") and `_` (underscore). They must begin
+with an alphabetic character or an underscore. The notational
+convention for `'` is to indicate a related definition, while
+underscore is mostly used to separate words or as a catch all for
+characters we'd like to use but are
+forbidden. [Camel case](https://en.wikipedia.org/wiki/Camel_case) is
+often used when other naming constraints aren't mandated.
+
+```
+mask = 7
+fooBar = 15
+fooBar' = fooBar && mask
+```
+
+Technically, Cryptol supports
+[Unicode](https://en.wikipedia.org/wiki/Unicode), but for simplicity I
+am pretending that it doesn't.
+
+
+Operators
+---------
+
+Cryptol's `:help` command will provide a brief description of the
+operators in this section by issuing `:help ` followed
+by the name of the operator in parenetheses. For example: `:help (@)`
+
+Many languages differentiate signed and unsigned numbers at the type
+level (e.g. C's `uint32` and `int32`). Cryptol has separate operators
+for signed operations which are indicated by a suffixed `$`. Most of
+the time you don't need them as cryptography tends to use nonnegative
+numbers.
+
+Following are some really quick examples of operators to remind you
+and show some tricks of Cryptol.
+
+### Arithmetic: `+`, `-`, `*`, `/`, `%` and `^^`
+#### Signed versions: `/$` and `%$`
+
+```sh
+labs::LanguageBasics::LanguageBasics> 1 + 1
+2
+labs::LanguageBasics::LanguageBasics> 1 + 1 : [1]
+0x0
+labs::LanguageBasics::LanguageBasics> 2^^127 - 1 // a 33 digit Mersenne prime
+170141183460469231731687303715884105727
+```
+
+### Bitwise logical: `~`, `&&`, `||` and `^`
+
+```sh
+labs::LanguageBasics::LanguageBasics> ~0b000011001101 && 0o4115 || 0x0d0 ^ 9
+0x8d9
+```
+
+Notice there's `0b...` for binary, `0o...` for octal and `0x...` for hexadecimal.
+
+### Comparison:`==`, `!=`, `<` , `<=`, `>` and `>=`
+#### Signed versions: `<$`, `<=$`, `>$` and `>=$`
+
+```sh
+labs::LanguageBasics::LanguageBasics> [~1, 1] == [6, 4 + 5]
+True
+labs::LanguageBasics::LanguageBasics> [~1, 1] == [6, 0b0100 + 5]
+False
+```
+
+Cryptol figures that the `6` in the first example requires three bits
+and that's the widest thing so both sides are of type `[2][3]` (two
+elements of three bits each). So `[~1, 1] == [6, 1] == [6, 4 + 5]`.
+
+The `0b0100` in the second example needs four bits, so both sides have
+type `[2][4]`. In this case `[~1, 1] == [14, 1]` while `[6, 4 + 5] ==
+[6, 9]` so equality fails.
+
+_**It is important to be precise about the widths of things!**_
```

Comment:
This appears to be the introduction of bit-level width granularity in Cryptol.  That feels like it should be a section in itself.

There are some non-intuitive (for me) aspects here.  Why does [6, 4+5] become [2][3] and not [2][4]?  Does the first item in the sequence decree the item widths?  I have soooo many questions!

Reply:
will address.

Reply:
Addressed by adding a ## Data section and reworking the explanation.

Reply:
Added a Data section and reworked this explanation. I think that address Zig's comment.